vllm/model_executor/parameter.py

```diff
@@ -0,0 +1,160 @@
+                "or column shard splitting, assume the weight is the same "
+                "for all partitions.")
+        self._use_metadata_loading = False
+
+    @property
+    def input_dim(self):
+        return self._input_dim
+
+    @property
+    def output_dim(self):
+        return self._output_dim
+
+    @property
+    def weight_loader(self):
+        return self._weight_loader
+
+    @property
+    def is_packed(self):
+        return self._is_packed
+
+    @property
+    def use_col_shard_split(self):
+        return self._use_col_shard_split
+
+    @property
+    def use_column_loading(self):
+        return self._use_col_loading
+
+    @property
+    def use_row_loading(self):
+        return self._use_row_loading
+
+    # TODO: should be part of ScalerToArrayvLLMParameter logic?
+    @property
+    def use_metadata_loading(self):
+        return self._use_metadata_loading
+
+    @use_metadata_loading.setter
+    def use_metadata_loading(self, value: bool):
+        self._use_metadata_loading = value
+
+
+class ScalerToArrayvLLMParameter(vLLMParameter):
+
+    def __init__(self, logical_widths: List[int], **kwargs):
+        self.logical_widths = logical_widths
+        self.qkv_idxs = {"q": 0, "k": 1, "v": 2}
+
+        super().__init__(**kwargs, use_col_shard_split=True)
+
+        if self.use_column_loading:
+            raise ValueError("Can only use one of column shard splitting "
+                             "or column default loading")
+
+    def _shard_id_as_int(self, shard_id: Union[str, int]) -> int:
+        if isinstance(shard_id, int):
+            return shard_id
+
+        assert isinstance(shard_id, str)
+        assert shard_id in self.qkv_idxs
+        return self.qkv_idxs[shard_id]
+
+    def col_shard_splitter(
+            self, param_data: torch.Tensor, loaded_weight: torch.Tensor,
+            shard_id: Union[str, int]) -> Tuple[torch.Tensor, torch.Tensor]:
+        shard_id = self._shard_id_as_int(shard_id)
+        offset = sum(self.logical_widths[:shard_id])
+        size = self.logical_widths[shard_id]
+        # update loaded weight with copies for broadcast.
+        loaded_weight = loaded_weight.repeat(size)
+        return param_data[offset:offset + size], loaded_weight
+
+
+class PackedvLLMParameter(vLLMParameter):
+
+    def __init__(self,
+                 packed_factor: int,
+                 packed_dim: int,
+                 marlin_tile_size: Optional[int] = None,
+                 **kwargs):
+        self._packed_factor = packed_factor
+        self._packed_dim = packed_dim
+        self._marlin_tile = marlin_tile_size
+        super().__init__(**kwargs, packed=True)
+
+    @property
+    def packed_dim(self):
+        return self._packed_dim
+
+    @property
+    def packed_factor(self):
+        return self._packed_factor
+
+    @property
+    def marlin_tile(self):
+        return self._marlin_tile
+
+    def _adjust_marlin_shard(self, shard_size, shard_offset):
```

Comment:
I think we should rename these:
- `_adjust_shard_indexes_for_marlin`
- `_adjust_shard_indexes_for_packing`